packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1002 @@
+      final VoidCallback listener = () {
+        if (bucket == null)
+          return;
+        _updateProperty(property);
+      };
+      property.addListener(listener);
+      _properties[property] = listener;
+    }
+
+    assert(
+      property._id == id &&
+      property._owner == this &&
+      _properties.containsKey(property)
+    );
+
+    property.initWithValue(initialValue);
+    if (!hasSerializedValue && property.enabled && bucket != null) {
+      _updateProperty(property);
+    }
+
+    assert(() {
+      _debugPropertiesWaitingForReregistration?.remove(property);
+      return true;
+    }());
+  }
+
+  /// Unregisters a [RestorableProperty] from state restoration.
+  ///
+  /// The value of the `property` is removed from the restoration data and it
+  /// will not be restored if that data is used in a future state restoration.
+  ///
+  /// Calling this method is uncommon, but may be necessary if the data of a
+  /// [RestorableProperty] is only relevant when the [State] object is in a
+  /// certain state. When the data of a property is no longer necessary to
+  /// restore the internal state of a [State] object, it may be removed from the
+  /// restoration data by calling this method.
+  @protected
+  void unregisterFromRestoration<T>(RestorableProperty<T> property) {
+    assert(property != null);
+    assert(property._owner == this);
+    _bucket?.remove<Object>(property._id);
+    _unregister(property);
+  }
+
+  /// Must be called when the value returned by [restorationId] changes.
+  ///
+  /// This method is automatically called from [didUpdateWidget]. Therefore,
+  /// manually invoking this method may be omitted when the change in
+  /// [restorationId] was caused by an updated widget.
+  @protected
+  void didUpdateRestorationId() {
+    if (_bucket?.id != restorationId && !restorePending) {
+      _updateBucketIfNecessary();
+    }
+  }
+
+  @override
+  void didUpdateWidget(S oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    didUpdateRestorationId();
+  }
+
+  /// Whether [restoreState] will be called at the beginning of the next build
+  /// phase.
+  ///
+  /// Returns true when new restoration data has been provided to the mixin, but
+  /// the registered [RestorableProperty]s have not been restored to their new
+  /// values (as described by the new restoration data) yet. The properties will
+  /// get the values restored when [restoreState] is invoked at the beginning of
+  /// the next build cycle.
+  ///
+  /// While this is true, [bucket] will also still return the old bucket with
+  /// the old restoration data. It will update to the new bucket with the new
+  /// data just before [restoreState] is invoked.
+  bool get restorePending => _restorePending;
+  bool _restorePending = true;
+
+  List<RestorableProperty<Object>> _debugPropertiesWaitingForReregistration;
+  bool get _debugDoingRestore => _debugPropertiesWaitingForReregistration != null;
+
+  @override
+  void didChangeDependencies() {
```

Comment:
would implementing didUpdateWidget help for the finalizer stuff above?

Reply:
I may be missing something, but wouldn't that let you resolve issues where a widget is moving from one part of the tree to another and transferring ownership of a bucket?

Reply:
From reading some more, I think it would not help. Disregard.